File: BOJ/python3/1654.py
'''
랜선 자르기
https://www.acmicpc.net/problem/1654
'''
import sys
import time
start = time.time()  # 시작 시간
sys.stdin = open("input.txt", "r")  # input.txt 파일로 입력

wires = []

K, N = map(int, input().split())
for _ in range(K):
    wire = int(sys.stdin.readline())
    wires.append(wire)

ans = 0
l, r = 1, max(wires)
while l <= r:
    mid = (l+r) // 2

    cnt = 0
    for wire in wires:
        cnt = cnt + wire//mid

    if cnt >= N and mid > ans:
        ans = mid

    if cnt < N:
        r = mid - 1
    else:
        l = mid + 1

print(ans)
# 현재시각 - 시작시간 = 실행 시간(sec)
print("--------------------")
print(f"Elapsed Time: {round(time.time() - start, 3)}s")

File: BOJ/nodejs/1654.js
/*
랜선 자르기
https://boj.kr/1436
 */
const start = new Date().getTime();

const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .split("\n");

const [K, N] = input[0].split(" ").map(Number);
const cables = new Array(K);
for (let k = 0; k < K; k++) {
  cables[k] = 1 * input[1 + k];
}

let ans = 0;
let [left, right] = [1, Math.max(...cables)];
let mid, cnt;
while (left <= right) {
  mid = Math.floor((left + right) / 2);
  cnt = cables
    .map((cable) => Math.floor(cable / mid))
    .reduce((acc, cur) => acc + cur, 0);
  if (cnt >= N) {
    ans = Math.max(ans, mid);
  }

  cnt >= N ? (left = mid + 1) : (right = mid - 1);
}

console.log(ans);

const end = new Date().getTime();
console.log("--------------------");
console.log(`Elapsed Time: ${(new Date().getTime() - start) / 1000}s`);
